csharp/src/Apache.Arrow/C/CArrowSchemaImporter.cs

```diff
@@ -0,0 +1,285 @@
+        /// CArrowSchema.Free(importedPtr);
+        /// </code>
+        /// </examples>
+        public static unsafe ArrowType ImportType(CArrowSchema* ptr)
+        {
+            using var importedType = new ImportedArrowSchema(ptr);
+            return importedType.GetAsType();
+        }
+
+        /// <summary>
+        /// Import C pointer as an <see cref="Field"/>.
+        /// </summary>
+        /// <examples>
+        /// Typically, you will allocate a uninitialized CArrowSchema pointer,
+        /// pass that to external function, and then use this method to import
+        /// the result.
+        /// 
+        /// <code>
+        /// CArrowSchema* importedPtr = CArrowSchema.New();
+        /// foreign_export_function(importedPtr);
+        /// Field importedField = CArrowSchemaImporter.ImportField(importedPtr);
+        /// CArrowSchema.Free(importedPtr);
+        /// </code>
+        /// </examples>
+        public static unsafe Field ImportField(CArrowSchema* ptr)
+        {
+            using var importedField = new ImportedArrowSchema(ptr);
+            return importedField.GetAsField();
+        }
+
+        /// <summary>
+        /// Import C pointer as an <see cref="Schema"/>.
+        /// </summary>
+        /// <examples>
+        /// Typically, you will allocate a uninitialized CArrowSchema pointer,
+        /// pass that to external function, and then use this method to import
+        /// the result.
+        /// 
+        /// <code>
+        /// CArrowSchema* importedPtr = CArrowSchema.New();
+        /// foreign_export_function(importedPtr);
+        /// Field importedSchema = CArrowSchemaImporter.ImportSchema(importedPtr);
+        /// CArrowSchema.Free(importedPtr);
+        /// </code>
+        /// </examples>
+        public static unsafe Schema ImportSchema(CArrowSchema* ptr)
+        {
+            using var importedSchema = new ImportedArrowSchema(ptr);
+            return importedSchema.GetAsSchema();
+        }
+
+        private sealed unsafe class ImportedArrowSchema : IDisposable
+        {
+            private readonly CArrowSchema* _data;
+            private readonly bool _isRoot;
+
+            public ImportedArrowSchema(CArrowSchema* handle)
+            {
+                _data = (CArrowSchema*)handle;
+                if (_data->release == null)
+                {
+                    throw new ArgumentException("Tried to import a schema that has already been released.");
+                }
+                _isRoot = true;
+            }
+
+            public ImportedArrowSchema(CArrowSchema* handle, bool isRoot) : this(handle)
+            {
+                _isRoot = isRoot;
+            }
+
+            public void Dispose()
+            {
+                // We only call release on a root-level schema, not child ones.
+                if (_isRoot && _data->release != null)
+                {
+                    _data->release(_data);
```

Comment:
I don't think we should be releasing anything that we don't own. `_data` was passed by the public API, right? We don't own it, we shouldn't dispose of it.

Reply:
Is there a way in C# to indicate that something is meant to take ownership of a pointer? TBH I think I would prefer that over relying on the user to remember to call the release callback in the appropriate situation.

For reference, that's how the C++ import functions work:

https://github.com/apache/arrow/blob/5b2fbade23eda9bc95b1e3854b19efff177cd0bd/cpp/src/arrow/c/bridge.h#L86-L94

Reply:
I've added a remark to these methods to indicate the release callback will be called. And I've added a unit test to make sure that it is called, even if the import fails.

Reply:
Can you show me in the code where `release` is called? I'm not seeing the code for it. (Yes, I do see the comment, but I don't see the backing code, nor the place in https://arrow.apache.org/docs/format/CDataInterface.html where it says that when you "import" a C Data schema object, the incoming pointer is released.

I feel this is a bad pattern - releasing resources that you don't own/didn't create. It would be really surprising to me that I couldn't get a `CArrowSchema*` from somewhere, and I couldn't call `CArrowSchemaImporter.ImportSchema` multiple times on it.

cc @pitrou who has the "git blame" on the above bridge.h file.

Reply:
> I feel this is a bad pattern - releasing resources that you don't own/didn't create.

To be clear: releasing doesn't necessarily mean freeing memory. It could also mean decrementing a reference count on a shared pointer. The release callback is just a way for the consumer to inform the producer that it is no longer using those resources. If the producer also isn't using those resources, it will free them. That's commonly the case with `ArrowSchema`, since it's all just small temporary values we created for exchange. But in other cases it might keep them around. That will be the case with `ArrowArray`, where the buffers will be behind reference counted pointers that the producer might still have references to.

> It would be really surprising to me that I couldn't get a CArrowSchema* from somewhere, and I couldn't call CArrowSchemaImporter.ImportSchema multiple times on it.

I think it's basically just a trade off. Either you can only import once, and the release callback is handled for you. Or you are allowed to import multiple times, but you are responsible for also calling the release callback (or else risking memory leaks). It seems the current design expects that most situations people only care to import something once, and that it's more important to prevent accidental memory leaks than it is support multiple imports.

> Can you show me in the code where release is called?

Indeed it's a little complicated:

 1. When you import, you create a `SchemaImporter`: https://github.com/apache/arrow/blob/08fb86107b001e2e1903ae0e416ff51386d4239e/cpp/src/arrow/c/bridge.cc#L1212-L1213
 2. When the schema importer is passed the `ArrowSchema`, it passes that to an `ExportGuard` https://github.com/apache/arrow/blob/08fb86107b001e2e1903ae0e416ff51386d4239e/cpp/src/arrow/c/bridge.cc#L831
 3. When the `SchemaImporter` destructor is called, the `ExportGuard` destructor is called, and that calls the release callback https://github.com/apache/arrow/blob/08fb86107b001e2e1903ae0e416ff51386d4239e/cpp/src/arrow/c/util_internal.h#L61

Reply:
> I feel this is a bad pattern - releasing resources that you don't own/didn't create

This is the entire point. Someone (the producer) decided to share those resources with you, and they need to know when you are finished using those resources.

As @wjones127 said, the release callback will generally not be a straight deallocation; instead, it may decrement a dedicated reference count (or unpin the backing object if the producer is implemented in a GC-based language).

Reply:
I have 2 questions about this:

> they need to know when you are finished using those resources.

1. Isn't the fact that the function call returned to the caller enough to know that I'm finished using those resources?

>  it may decrement a dedicated reference count 

2. How do I "increment a dedicated reference count"? For example, if I received a `CArrowSchema*` from somewhere, and I need to pass it into a C# Importer and a C++ importer method. How could I tell the `CArrowSchema*` that it needs to allow `release` to be called twice?

Reply:
> Isn't the fact that the function call returned to the caller enough to know that I'm finished using those resources?

I think this will make more sense when you look at array import rather than schema import. The main point of the C data interface is to allow one implementation (the producer) to share array buffers with another implementation (the consumer). If the lifetime of those buffers was limited to the import function call, then that means the consumer would either be limited to using the arrays during the import or else have to copy the buffers (at which point you might as well use IPC).

As an example, imagine we used the C data interface to create bindings to DataFusion to run queries. 

```csharp
RecordBatch batch = ... 
DataFusionContext context;

// Uses C data interface to import batch (as a struct array)
context.ImportBatch("my_table", batch);

// This query is applied to the imported array
var result = context.Sql("SELECT sum(x * y) FROM my_table");
```

Here, if we assumed it was safe to cleanup the batch after calling `ImportBatch`, then we couldn't call the subsequent query to do any computation on it. We need those buffers to live as long as the consumer still needs them.

Reply:
> How do I "increment a dedicated reference count"? For example, if I received a CArrowSchema* from somewhere, and I need to pass it into a C# Importer and a C++ importer method. How could I tell the CArrowSchema* that it needs to allow release to be called twice?

I don't think you can import a `CArrowSchema*` more than once for that reason. So either the original producer would have to produce a `CArrowSchema` for each consumer, or else you could chain them, so that the C# importer imports the original one, and then the C# exporter creates a `CArrowSchema` that could be imported by C++.